src/include/denc.h

```diff
@@ -1845,11 +1846,26 @@ inline std::enable_if_t<traits::supported && !traits::featured> decode_nohead(
 // Helpers for versioning the encoding.  These correspond to the
 // {ENCODE,DECODE}_{START,FINISH} macros.
 
+// DENC_START interface suggests it is checking compatibility,
+// but the feature was unimplemented until SQUID.
+// Due to -2 compatibility rule we cannot bump up compat until U____ release.
+
 #define DENC_START(v, compat, p)					\
   __u8 struct_v = v;							\
   __u8 struct_compat = compat;						\
   char *_denc_pchar;							\
   uint32_t _denc_u32;							\
+  static_assert(CEPH_RELEASE >= 21 || compat == 1);			\
```

Comment:
I think it's a pretty reasonable compromise – although users are made aware about the problem a bit later, at compile time, the code churn is minimal. :+1: